File: WdtcCore/src/main/java/org/wdt/wdtc/core/game/Launcher.java
package org.wdt.wdtc.core.game;


import lombok.Getter;
import org.wdt.utils.io.FileUtils;
import org.wdt.utils.io.IOUtils;
import org.wdt.wdtc.core.auth.accounts.Accounts;
import org.wdt.wdtc.core.download.fabric.FabricDonwloadInfo;
import org.wdt.wdtc.core.download.forge.ForgeDownloadInfo;
import org.wdt.wdtc.core.download.quilt.QuiltDownloadInfo;
import org.wdt.wdtc.core.download.quilt.QuiltInstallTask;
import org.wdt.wdtc.core.game.config.GameConfig;
import org.wdt.wdtc.core.game.config.VersionInfo;
import org.wdt.wdtc.core.manger.GameDirectoryManger;
import org.wdt.wdtc.core.manger.GameFileManger;
import org.wdt.wdtc.core.manger.SettingManger;
import org.wdt.wdtc.core.utils.ModUtils;

import java.io.File;
import java.io.IOException;
import java.util.Objects;

@Getter
public class Launcher extends GameFileManger {
  private FabricDonwloadInfo FabricModInstallInfo;
  private ModUtils.KindOfMod kind = ModUtils.KindOfMod.Original;
  private ForgeDownloadInfo ForgeModDownloadInfo;
  private QuiltDownloadInfo QuiltModDownloadInfo;

  public Launcher(String version) {
    this(version, SettingManger.getSetting().getDefaultGamePath());
  }

  public Launcher(String version, File here) {
    super(version, here);
  }

  public static Launcher getPreferredLauncher() {
    SettingManger.Setting setting = SettingManger.getSetting();
    return setting.getPreferredVersion() != null
        ? ModUtils.getModTask(new Launcher(setting.getPreferredVersion()))
        : null;
  }

  public void setQuiltModDownloadInfo(QuiltInstallTask quiltModDownloadInfo) {
    kind = ModUtils.KindOfMod.QUILT;
    QuiltModDownloadInfo = quiltModDownloadInfo;
  }

  public void setFabricModInstallInfo(FabricDonwloadInfo fabricModInstallInfo) {
    kind = ModUtils.KindOfMod.FABRIC;
    this.FabricModInstallInfo = fabricModInstallInfo;
  }

  public void setKind(ModUtils.KindOfMod kind) {
    this.kind = kind;
  }


  public void setForgeModDownloadInfo(ForgeDownloadInfo forgeModDownloadInfo) {
    kind = ModUtils.KindOfMod.FORGE;
    ForgeModDownloadInfo = forgeModDownloadInfo;
  }

  public GameConfig getGameConfig() {
    return new GameConfig(this);
  }

  public void beforLaunchTask() throws IOException {
    if (SettingManger.getSetting().isChineseLanguage() && FileUtils.isFileNotExists(getGameOptionsFile())) {
      String Options = IOUtils.toString(Objects.requireNonNull(getClass().getResourceAsStream("/assets/options.txt")));
      FileUtils.writeStringToFile(getGameOptionsFile(), Options);
    }
  }

  public void CleanKind() {
    kind = ModUtils.KindOfMod.Original;
  }

  public Accounts getAccounts() {
    return new Accounts();
  }


  public VersionInfo getVersionInfo() {
    return new VersionInfo(this);
  }

  public GameDirectoryManger getGameDirectoryManger() {
    return this;
  }


  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (!(o instanceof Launcher launcher)) return false;
    if (launcher.getVersionNumber().equals(VersionNumber)) {
      return launcher.getKind() == kind;
    }
    return false;
  }

  @Override
  public int hashCode() {
    return Objects.hash(VersionNumber, kind);
  }

  @Override
  public String toString() {
    return "Launcher(version=" + VersionNumber + ", kind=" + kind + ")";


  }
}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/game/Launcher.kt
package org.wdt.wdtc.core.game

import org.wdt.utils.io.FileUtils
import org.wdt.utils.io.IOUtils
import org.wdt.utils.io.writeStringToFile
import org.wdt.wdtc.core.download.fabric.FabricDonwloadInfo
import org.wdt.wdtc.core.download.forge.ForgeDownloadInfo
import org.wdt.wdtc.core.download.quilt.QuiltDownloadInfo
import org.wdt.wdtc.core.manger.GameFileManger
import org.wdt.wdtc.core.manger.setting
import org.wdt.wdtc.core.utils.KindOfMod
import org.wdt.wdtc.core.utils.setModTask
import java.io.File
import java.io.IOException
import java.util.*

class Launcher @JvmOverloads constructor(version: String, here: File = setting.defaultGamePath) :
  GameFileManger(version, here) {

  var fabricModInstallInfo: FabricDonwloadInfo? = null
    set(value) {
      kind = KindOfMod.FABRIC
      field = value
    }

  var forgeModDownloadInfo: ForgeDownloadInfo? = null
    set(value) {
      kind = KindOfMod.FORGE
      field = value
    }


  var quiltModDownloadInfo: QuiltDownloadInfo? = null
    set(value) {
      kind = KindOfMod.FABRIC
      field = value
    }


  var kind = KindOfMod.ORIGINAL


  @Throws(IOException::class)
  fun beforLaunchTask() {
    if (setting.chineseLanguage && FileUtils.isFileNotExists(gameOptionsFile)) {
      val options = IOUtils.toString(Objects.requireNonNull(javaClass.getResourceAsStream("/assets/options.txt")))
      gameOptionsFile.writeStringToFile(options)
    }
  }

  fun cleanKind() {
    kind = KindOfMod.ORIGINAL
  }


  override fun toString(): String {
    return "Launcher(versionNumber=$versionNumber, kind=$kind)"
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (javaClass != other?.javaClass) return false

    other as Launcher

    if (versionNumber != other.versionNumber) return false
    if (kind != other.kind) return false
    if (fabricModInstallInfo != other.fabricModInstallInfo) return false
    if (forgeModDownloadInfo != other.forgeModDownloadInfo) return false
    if (quiltModDownloadInfo != other.quiltModDownloadInfo) return false

    return true
  }

  override fun hashCode(): Int {
    var result = fabricModInstallInfo?.hashCode() ?: 0
    result = 31 * result + (forgeModDownloadInfo?.hashCode() ?: 0)
    result = 31 * result + (quiltModDownloadInfo?.hashCode() ?: 0)
    result = 31 * result + kind.hashCode()
    result = 31 * result + versionNumber.hashCode()
    return result
  }

}

val preferredLauncher: Launcher?
  get() = if (setting.preferredVersion != null) Launcher(setting.preferredVersion!!).setModTask() else null
